### PR TITLE
Pretty-print dataclasses

### PR DESCRIPTION
**Pull request by francescocastelli**
Unfortunately the built-in pprint module support pretty-print of dataclasses only from python 3.10. The code that I wrote in method `__str__` of OpInfo should do the same job and should also work for any dataclass. For now I've put it there but we can create a function and put it somewhere where is accessible also for other dataclasses. Also the max width (80) is now hardcode but it would ideally be the parameter of the function.

when you call print on an OpInfo you get: 
```
OpInfo(name = '__getitem__',
       ref = None,
       aliases = (),
       variant_test_name = '',
       op = <slot wrapper '__getitem__' of 'torch._C._TensorBase' objects>,
       method_variant = <slot wrapper '__getitem__' of 'torch._C._TensorBase' objects>,
       inplace_variant = None,
       skips = (<torch.testing._internal.common_methods_invocations.DecorateInfo object at 0x7f463acbca90>,
                <torch.testing._internal.common_methods_invocations.DecorateInfo object at 0x7f463acbcae0>),
       decorators = (<torch.testing._internal.common_methods_invocations.DecorateInfo object at 0x7f463acbca90>,
                     <torch.testing._internal.common_methods_invocations.DecorateInfo object at 0x7f463acbcae0>),
       sample_inputs_func = <function sample_inputs_getitem at 0x7f463acc6af0>,
       reference_inputs_func = None,
       error_inputs_func = None,
       sample_inputs_sparse_coo_func = <function _DecoratorContextManager.__call__.<locals>.decorate_context at 0x7f463acc6b80>,
       sample_inputs_sparse_csr_func = <function _DecoratorContextManager.__call__.<locals>.decorate_context at 0x7f463acc6c10>,
       dtypes = {torch.int16,
                 torch.float64,
                 torch.int32,
                 torch.int64,
                 torch.complex64,
                 torch.float16,
                 torch.bfloat16,
                 torch.uint8,
                 torch.complex128,
                 torch.bool,
                 torch.float32,
                 torch.int8},
       dtypesIfCUDA = {torch.int16,
                       torch.float64,
                       torch.int32,
                       torch.int64,
                       torch.complex64,
                       torch.float16,
                       torch.bfloat16,
                       torch.uint8,
                       torch.complex128,
                       torch.bool,
                       torch.float32,
                       torch.int8},
       dtypesIfROCM = {torch.int16,
                       torch.float64,
                       torch.int32,
                       torch.int64,
                       torch.complex64,
                       torch.float16,
                       torch.bfloat16,
                       torch.uint8,
                       torch.complex128,
                       torch.bool,
                       torch.float32,
                       torch.int8},
       backward_dtypes = {torch.int16,
                          torch.float64,
                          torch.int32,
                          torch.int64,
                          torch.complex64,
                          torch.float16,
                          torch.bfloat16,
                          torch.uint8,
                          torch.complex128,
                          torch.bool,
                          torch.float32,
                          torch.int8},
       backward_dtypesIfCUDA = {torch.int16,
                                torch.float64,
                                torch.int32,
                                torch.int64,
                                torch.complex64,
                                torch.float16,
                                torch.bfloat16,
                                torch.uint8,
                                torch.complex128,
                                torch.bool,
                                torch.float32,
                                torch.int8},
       backward_dtypesIfROCM = {torch.int16,
                                torch.float64,
                                torch.int32,
                                torch.int64,
                                torch.complex64,
                                torch.float16,
                                torch.bfloat16,
                                torch.uint8,
                                torch.complex128,
                                torch.bool,
                                torch.float32,
                                torch.int8},
       supports_out = False,
       supports_autograd = True,
       supports_gradgrad = True,
       supports_fwgrad_bwgrad = True,
       supports_inplace_autograd = False,
       supports_forward_ad = True,
       gradcheck_wrapper = <function OpInfo.<lambda> at 0x7f463a7a40d0>,
       check_batched_grad = True,
       check_batched_gradgrad = True,
       check_batched_forward_grad = True,
       check_inplace_batched_forward_grad = True,
       gradcheck_nondet_tol = 0.0,
       gradcheck_fast_mode = None,
       aten_name = '__getitem__',
       decomp_aten_name = None,
       aten_backward_name = None,
       assert_autodiffed = False,
       autodiff_nonfusible_nodes = ['aten::__getitem__'],
       autodiff_fusible_nodes = [],
       supports_sparse = False,
       supports_scripting = False,
       supports_sparse_csr = False,
       test_conjugated_samples = True,
       test_neg_view = True,
       assert_jit_shape_analysis = False,
       supports_expanded_weight = False)
```

cc @ezyang

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -895,6 +895,26 @@ def __call__(self, *args, **kwargs):
         """Calls the function variant of the operator."""
         return self.op(*args, **kwargs)
 
+    def __str__(self):
+        class_name = self.__class__.__name__
+        indent = (len(class_name) + 1) * " "
```
Comment (ezyang) [inline torch/testing/_internal/common_methods_invocations.py:900]: Some of the OpInfo class names are quite long; maybe we should just newline and indent a fixed amount (2 spaces, perhaps?)

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -895,6 +895,26 @@ def __call__(self, *args, **kwargs):
...
+        field_list = [(f.name, getattr(self, f.name)) for f in fields(self) if f.repr]
+
+        fields_str = []
+        for f in field_list:
+            field_str = f"{f[0]} = {repr(f[1])}"
+
+            # 80 is the max_width, if the len is over it we split the args over multiple lines
+            if len(field_str) >= 80:
+                max_indent = (len(f[0]) + 3) * " "
+                field_str = f",\n{indent + max_indent}".join(field_str.split(","))
```
Comment (ezyang) [inline torch/testing/_internal/common_methods_invocations.py:911]: This won't do the right thing if there is a nested dataclass right?

**Comment by ezyang** [comment]
> The problem now is when you have a list of dataclasses (for instance) as member. I will look into that 

Can we come up with a unified strategy for dealing with this? Even if that strategy is "use pprint in Python 3.10, so it doesn't matter how hacky this code is"

**Comment by francescocastelli** [comment]
Yes, that strategy can be used from the beginning (if python >= 3.10 we use pprint otherwise fall back to this code). With python< 3.10 handling all possible cases would make the code even "hackier", so I'm trying to think of a cleaner solution. If you have any suggestions let me know :)

**Comment by ezyang** [comment]
I guess my point is, if the PoR is "when 3.10 shows up we delete this code" then I don't care how hacky the code is, as long as it works for the use cases I'll ship it.

**Comment by francescocastelli** [comment]
Where should we place this function such that is accessible also in other places? I've seen files like torch/_utils.py or torch/_utils_internal.py but don't know if these are the right places.

**Comment by ezyang** [comment]
Check it in _utils.

**Comment by francescocastelli** [comment]
This function should produce the same result as pprint when python < 3.10. I put the indentation back to the length of the class name to be consistent with pprint (because it always does that, the indent parameter only controls indentation at nested levels). 

I've tried in torchgen, this is the result when printing a NativeFunction with python < 3.10:

```
NativeFunction(func=FunctionSchema(name=OperatorName(name=BaseOperatorName(base='log_sigmoid_forward',
                                                                           inplace=False,
                                                                           dunder_method=False),
                                                     overload_name=''),
                                   arguments=Arguments(pre_self_positional=(),
                                                       self_arg=SelfArgument(argument=Argument(name='self',
                                                                                               type=BaseType(name=<BaseTy.Tensor: 3>),
                                                                                               default=None,
                                                                                               annotation=None)),
                                                       post_self_positional=(),
                                                       pre_tensor_options_kwarg_only=(),
                                                       tensor_options=None,
                                                       post_tensor_options_kwarg_only=(),
                                                       out=()),
                                   returns=(Return(name='output',
                                                   type=BaseType(name=<BaseTy.Tensor: 3>),
                                                   annotation=None), 
                                            Return(name='buffer',
                                                   type=BaseType(name=<BaseTy.Tensor: 3>),
                                                   annotation=None))),
               use_const_ref_for_mutable_tensors=False,
               device_guard=True,
               device_check=<DeviceCheckType.NoCheck: 0>,
               python_module='nn',
               category_override=None,
               variants=(<Variant.function: 1>),
               manual_kernel_registration=False,
               manual_cpp_binding=False,
               loc=Location(file='/home/francesco/projects/pytorch/cmake/../aten/src/ATen/native/native_functions.yaml',
                            line=9200),
               ufunc_inner_loop={},
               structured=False,
               structured_delegate=None,
               structured_inherits=None,
               precomputed=None,
               cpp_no_default_args=(),
               is_abstract=True,
               has_composite_implicit_autograd_kernel=False,
               has_composite_explicit_autograd_kernel=False,
               tags=())
```


I still get an error (`ModuleNotFoundError: No module named 'torch.version`') though when trying to import torch/_utils.py in torchgen/gen.py.

**Comment by ezyang** [comment]
I have a better idea: let's put it in `torchgen` somewhere (maybe `torchgen/utils.py`), now that torchgen is part of the packaging I think it's OK to import it!!!

**Comment by francescocastelli** [comment]
Done! There are some problems with `import torchgen` in windows

**Comment by ezyang** [comment]
it's probably because we're not preserving the torchgen directory. looking...

**Comment by ezyang** [comment]
patch this in

```
diff --git a/.jenkins/pytorch/win-test-helpers/build_pytorch.bat b/.jenkins/pytorch/win-test-helpers/build_pytorch.bat
index 9d05763d0e..c3650856d4 100644
--- a/.jenkins/pytorch/win-test-helpers/build_pytorch.bat
+++ b/.jenkins/pytorch/win-test-helpers/build_pytorch.bat
@@ -141,7 +141,7 @@ python setup.py install --cmake && sccache --show-stats && (
   if "%BUILD_ENVIRONMENT%"=="" (
     echo NOTE: To run `import torch`, please make sure to activate the conda environment by running `call %CONDA_PARENT_DIR%\Miniconda3\Scripts\activate.bat %CONDA_PARENT_DIR%\Miniconda3` in Command Prompt before running Git Bash.
   ) else (
-    7z a %TMP_DIR_WIN%\%IMAGE_COMMIT_TAG%.7z %CONDA_PARENT_DIR%\Miniconda3\Lib\site-packages\torch %CONDA_PARENT_DIR%\Miniconda3\Lib\site-packages\caffe2 && copy /Y "%TMP_DIR_WIN%\%IMAGE_COMMIT_TAG%.7z" "%PYTORCH_FINAL_PACKAGE_DIR%\"
+    7z a %TMP_DIR_WIN%\%IMAGE_COMMIT_TAG%.7z %CONDA_PARENT_DIR%\Miniconda3\Lib\site-packages\torch %CONDA_PARENT_DIR%\Miniconda3\Lib\site-packages\torchgen %CONDA_PARENT_DIR%\Miniconda3\Lib\site-packages\caffe2 && copy /Y "%TMP_DIR_WIN%\%IMAGE_COMMIT_TAG%.7z" "%PYTORCH_FINAL_PACKAGE_DIR%\"
     if errorlevel 1 exit /b
     if not errorlevel 0 exit /b
 
```

**Comment by francescocastelli** [comment]
perfect done!

**Comment by mruberry** [comment]
Thanks @francescocastelli! Test framework-related changes look great, of course, so I'll let Ed review this from a torchgen perspective

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**